content/src/content/docs/learn/tutorials/hooking-up-a-database.mdx

```diff
@@ -0,0 +1,395 @@
+);
+
+const main = Effect.gen(function* () {
+  yield* Effect.promise(() => db.connect());
+  app.use(yield* NodeHttpServer.makeHandler(router));
+  app.listen(3000);
+  yield* Effect.never;
+});
+
+NodeRuntime.runMain(main);
+```
+
+We'll also need to add some new dependencies to our project:
+
+```shell
+bun add pg @types/pg
+```
+
+Then you can run the server with:
+
+```shell
+bun index.ts
+```
+
+And in another shell we can test the endpoint.
+
+```shell
+http get localhost:3000/users
+```
+
+```http
+HTTP/1.1 200 OK
+Content-Length: 57
+Content-Type: application/json; charset=utf-8
+Date: Fri, 14 Feb 2025 14:28:02 GMT
+ETag: W/"39-6Qu85qIU12mhgGYWwHErQVfJRxI"
+X-Powered-By: Express
+
+{
+    "users": [
+        {
+            "id": 1,
+            "name": "Alice"
+        },
+        {
+            "id": 2,
+            "name": "Bob"
+        }
+    ]
+}
+```
+
+## The direct approach
+
+If we were to copy what we did with the `/health` endpoint, we'd end up with
+code that looked like this:
+
+```ts twoslash
+import { HttpRouter, HttpServer, HttpServerResponse } from "@effect/platform"
+import { NodeHttpServer, NodeRuntime } from "@effect/platform-node"
+import { Effect, Layer } from "effect"
+import { createServer } from "node:http"
+import { Client } from "pg";
+const db = new Client({ user: "postgres" });
+const health = Effect.sync(() => {
+  return HttpServerResponse.text("ok")
+})
+//---cut---
+// @errors: 2379
+const users = Effect.promise(async () => {
+  const { rows } = await db.query("SELECT * FROM users");
+  return HttpServerResponse.json({ users: rows });
+});
+
+const router = HttpRouter.empty.pipe(
+  HttpRouter.get("/health", health),
+  HttpRouter.get("/users", users),
+)
+```
+
+This doesn't work, as we can see from the error. The reason is that
+`HttpServerResponse.json` is an operation that might fail. Not all JavaScript
+objects can be converted to JSON.
+
+```ts twoslash
+JSON.stringify({ big: 10n })
+//=> TypeError: JSON.stringify cannot serialize BigInt.
+```
+
+For this reason, `HttpServerResponse.json` returns an `Effect` that can fail.
+We didn't see this earlier because `HttpServerResponse.text` can't fail, and
+so just returns an `HttpServerResponse` directly.
+
+## Error handling in Effect
+
+When we introduced the `Effect` type in the previous section, we dealt
+exclusively with Effects that couldn't fail. That won't get us very far if
+we're building anything non-trivial.
+
+`Effect` is a generic type that takes 3 arguments:
+
+```typescript
+Effect<Success, Error, Requirements>
+```
+
+`Success` is the type the Effect returns when it succeeds, `Error` is the type
+it returns when it fails, and we'll talk about `Requirements` a bit later.
+
+When we created our example Effects in the previous section, these all took on
+the type `Effect<string, never, never>`.
+
+```ts twoslash
+import { Effect } from "effect"
+//---cut---
+Effect.sync(() => "Hello, world!")
+//=> Effect.Effect<string, never, never>
```

Comment:
Just as an FYI, if you want to use Twoslash to display the type of something you can use the `// ^?` query operator:

```suggestion
import { Effect } from "effect"
//---cut---
const _ = Effect.sync(() => "Hello, world!")
//        ^?
```

I see you're using it in the snippet below, so I assume this was deliberate 😄